### ISSUE TITLE
Change the returns in `FitttingWithOutlierRemoval`

### ISSUE THREAD

**Issue by nden** [astropy/astropy#7350]
Currently `FittingWithOutlierRemoval` returns `filtered_data, fitted_model`. For consistence with other fitters and to decrease memory usage It was suggested in #7199 that 

- the order of returns is changed - the fitted model should be the first returned variable
- it should return the mask, not the filtered data

**Comment by jehturner** [comment]
In addition to those considerations, it is noted in https://github.com/astropy/astropy/pull/7199#issuecomment-375811113 that the `filtered_data.data` values currently returned are incorrect for masked points. What should happen now is that `filtered_data` is a copy of the `y` or `z` input, but with a mask attached/updated. At each rejection iteration, however, model values get subtracted from `filtered_data` and then added back in, with points eliminated in between. Because the arithmetic is skipped for masked points, the newly-rejected ones never get their model values restored. I believe this only affects masked points, but it leads to some discontinuous artifacts that misrepresent what was fitted, eg. when plotting the results. (There might also be unwanted, numerical-precision differences from the input due to the repeated arithmetic.)

Returning only the mask will avoid this, because users will never see the bad data array copy. It's not difficult to avoid the "corruption" in the first place, by performing in-place operations on `data` instead of the masked array, but that's marginally more verbose and the bad values should be harmless internally (unless someone adds debugging code to plot intermediate results, like I did?).

**Comment by jehturner** [comment]
Are you working on this, @nden?

There's one very basic aspect of this algorithm that I'm unsure about off hand, @mirca @nden, which relates to my comment above. Currently, the rejection of points is cumulative from one iteration to the next, isn't it? However, it's quite possible that some points will appear deviant at an early iteration, due to the fit not being accurate yet, but then turn out to fall closer to the final fit. So should the algorithm in fact be restoring those points between iterations and doing the rejection from scratch? Perhaps the answer is that cumulative rejection converges correctly because of the way the standard deviation scales down between iterations? It looks like IRAF does the same thing from a quick experiment, but I'm not 100% sure.

**Comment by nden** [comment]
@jehturner I won't have time for this in the next two weeks (perhaps longer) so if you can/want please submit a PR. 

I think the last issue you raised may go in a different PR. I want to split the interface change in its own PR.
I'm also not sure about the cumulative rejection. It's not obvious to me that cumulative rejection is correct especially with non-linear fitters. Perhaps the RANSAC algorithm avoids this problem. @mirca do you have more info on this?

**Comment by jehturner** [comment]
If I understand the IRAF code correctly, it does the same thing (in `xtools/icfit/ic_deviantr`).

**Comment by nden** [comment]
Let's do it this way and document how it is done.

**Comment by jehturner** [comment]
I confirm virtually identical (not far off machine precision) results to IRAF when fitting image rows containing cosmic rays with `Legendre1D`, as long as I pass `iters=1` to `sigma_clip` when instantiating `FittingWithOutlierRemoval` (otherwise the sigma limits mean quite different thresholds in the 2 cases for the first 1-2 iterations, causing modeling to reject many more points than IRAF).

**Comment by jehturner** [comment]
Do we always want to return an array for the mask, even if there are no masked values (eg. when `niter=0`), or should it return `np.ma.nomask` in that case? In general, we might have to check the mask with `np.all` to determine whether the `outlier_func` clipped anything, so always returning an array might be easier and a simpler API.

**Comment by ysBach** [comment]
I agree with @jehturner that the mask (rejection) should _not_ be cumulative. 
But may I see an example which gives different result depending on the cumulativeness? I believe cumulativeness must give difference at some point, but I cannot find an example after testing many cases..

**Comment by jehturner** [comment]
I don't have a very good example, other than maybe in my head. It would be interesting to investigate this more carefully at some point, but for now, I believe I've established that the behaviour is the same as in IRAF, which is known to do something reasonable, so that's a good baseline. I think the difference due to masking cumulatively is generally less significant than one might expect, because where artifacts bias the fit, they also raise the RMS accordingly, such that only the most deviant values get excluded by the initial iteration(s).

**Comment by ysBach** [comment]
I am sorry if it is not favorable to discuss it in this thread.. 

After wrote the comment, I took some time to think about this issue by myself. It's my first time thinking about the cumulative rejection @jehturner pointed out.

I was a bit confused as I couldn't find any single example which gives different result depending on the cumulativeness of the rejection, and I had to find some reason to justify this result: "the list of rejected points is identical regardless of cumulativeness in rejection every time I tested". Since I cannot say "they are the same in all the cases" **I basically agree to modify the code to do the non-cumulative rejection, unless someone gives mathematical reason not to do so**. But to give some reason to my result, I made the following explanation to the situation (sorry I cannot give a mathematical proof on this...):

-----

If I am understanding it correctly, ``FittingWithOutlierRemoval``'s algorithm is like this:

1. *initial fitting* using ``fitter`` from our initial guess (``f_0``).
   *  I usually use ``fitter=LevMarLSQFitter()``.
2. calculate residuals by ``outlier_func``
   * I usually use ``outlier_func=sigma_clip()`` (``sigma`` and ``iters`` goes here)
3. using only the survived (non-rejected) points, do the fitting again using ``fitter`` to get ``f_i``.
4. go to 2 until ``i == niter``.

The "cumulativeness" occurs in the second step. But ``FittingWithOutlierRemoval`` does the initial fitting before the outliers are cumulated. So in my opinion, although I cannot mathematically prove it, *the rejected points have no hope to be resurrected once the initial fitting (1st step) is done, even though we don't do the cumulative rejection.* If the "resurrection" of a rejected point happens, that means the Lev-Mar algorithm had rejected a "good" point at a certain iteration, and that is merely the problem or weak point of Lev-Mar fitter. (If the true function is ``f_t``, then I guess the outlier rejection should occur in a way that ``f_i`` gets closer to ``f_t`` as ``i`` increases. I guess Lev-Mar algorithm is developed in a way that we can find ``f_t`` from ``f_0``which is reasonably close to ``f_t``.)

-----

But this is based on my experience, and no firm reason is there to support cumulative rejection. So at least I should say doing non-cumulative rejection is favorable (and safer) at this moment..

**Comment by nden** [comment]
Talking to modeling users at pyastro2018 it seems like everyone prefers to have this change done although it may break some code. So making a last call for expressing strong opposition before merging. @astrofrog @eteq

**Comment by jehturner** [comment]
Sorry about all the tangential discussion that I started here. As far as the API changes are concerned, you can ignore everything but the first 2 comments above. The rest of the discussion about whether masking should be cumulative between iterations should be addressed separately.
